### PR TITLE
change(web): move KeyboardProcessor.Codes into common/web/types 🏗️

### PR DESCRIPTION
**Pull request by ermshiperete**
Fixes: #8146 

@keymanapp-test-bot skip

### CODE REVIEW

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Thank you for this refactor PR. I am going to requesting a significant change here.

I would like to be careful about _what_ we move into common/types. I am already trying to extricate myself from Developer-specific files that I put there, so don't want to do the same thing for KeymanWeb-specific content. 

The constants in codes.ts do belong in common/types -- but they are nearly all already there, in kmx.ts:

https://github.com/keymanapp/keyman/blob/3497e97b00a730be0a18aa8f34cb65d61f767b2a/common/web/types/src/kmx/kmx.ts#L346-L364

and virtual-key-constants.ts:

https://github.com/keymanapp/keyman/blob/2b2661bb1bea711a13454ad568d00aefeddc8e4a/common/web/types/src/consts/virtual-key-constants.ts#L5-L143

Given this, I am not sure that this move makes sense -- we don't solve the WETness of the constants. So here's an alternate proposal:

1. consider a reorg of kmx.ts to move modifier constants into consts/modifier-key-constants.ts (and re-export them from kmx.ts to maintain 100% API surface consistency)
2. keep codes.ts in keyboard-processor
3. harmonize any inconsistencies between the constants in codes.ts and modifier-key-constants.ts / virtual-key-constants.ts. This may mean adding some constants to those two files (but without changing the pattern in which they are declared.)
4. finally, remove the constant declarations from codes.ts, and instead import and re-export the ones found in modifier-key-constants.ts / virtual-key-constants.ts. If necessary, add wrappers so that we can match the way kmw currently uses the constants. (Avoid rewriting other aspects of kmw at this point)


I think this would make the change quite a lot smaller, as most of the references in kmw wouldn't change, just codes.ts. It would also address the WETness, which IMO is the primary driver for this refactor.

I appreciate that this is a big change request -- but really feel like it will result in a better outcome.

File: web/src/engine/osk/src/index.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-export { Codes, DeviceSpec, Keyboard, KeyboardProperties, SpacebarText } from '@keymanapp/keyboard-processor';
+export { DeviceSpec, Keyboard, KeyboardProperties, SpacebarText } from '@keymanapp/keyboard-processor';
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/src/engine/osk/src/index.ts:1]: Considering that we only even modularized to this extent with Keyman 17, I severely doubt anyone external is relying upon `Codes` here.  If everything is building fine, we're good.

I would be a touch worried about Web's test-sequence "recording" module... but that's built via `./web/build.sh tools`, which is triggered during standard Web CI builds.  As even _that_ didn't fall over, in the absolute worst case, all that would be needed is to tidy up the relevant pages.

tl;dr: nah, no need to re-export.

File: common/web/keyboard-processor/src/index.ts [hunk 2]
```diff
@@ -24,8 +24,6 @@ export {
 export { default as SpacebarText } from "./keyboards/spacebarText.js";
 export { default as StateKeyMap } from "./keyboards/stateKeyMap.js";
 
-export { default as Codes } from "./text/codes.js";
-export * from "./text/codes.js";
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/index.ts:28]: Eh, if it builds, I assume it's fine.  Standard Web CI builds even build the development tools, which would be the only thing I'd worry about falling over.  (They'd fail the build if they fell over.)

**Comment by jahorton** [comment]
> [...]
>
> 3. harmonize any inconsistencies between the constants in codes.ts and modifier-key-constants.ts / virtual-key-constants.ts. This may mean adding some constants to those two files (but without changing the pattern in which they are declared.)
> 
> 4. finally, remove the constant declarations from codes.ts, and instead import and re-export the ones found in modifier-key-constants.ts / virtual-key-constants.ts. If necessary, add wrappers so that we can match the way kmw currently uses the constants. (Avoid rewriting other aspects of kmw at this point)

Why would we not just rework the definition patterns to be consistent for each Codes property?   The excerpted `USVirtualKeyCodes` pattern is an exact-match, save _maybe_ for some constant names.  Is there a particular reason not to rework the following block to use a similar format?  I think it'd be better for maintenance than having to assign each constant (as below / in `kmx`) to an object (as provided within `Codes`) as a separate step.

> https://github.com/keymanapp/keyman/blob/3497e97b00a730be0a18aa8f34cb65d61f767b2a/common/web/types/src/kmx/kmx.ts#L346-L364

I do understand we'd probably need a wrapper to "bundle" all the codes into their "final form" - Web's current `Codes` object format, with their current property names - that much makes sense.

**Comment by mcdurdin** [comment]
> Why would we not just rework the definition patterns to be consistent for each Codes property? The excerpted `USVirtualKeyCodes` pattern is an exact-match, save _maybe_ for some constant names. Is there a particular reason not to rework the following block to use a similar format? I think it'd be better for maintenance than having to assign each constant (as below / in `kmx`) to an object (as provided within `Codes`) as a separate step.

I would like to avoid a cascade of changes into Developer at this point -- there are many changes happening in Developer at the same time. That block is a match for kmx_file.h, also.

But we could certainly put an object wrapper into common/types (that is then imported into Codes.ts). Then at least the wrapper and the original definitions are in the same place.

**Comment by ermshiperete** [comment]
So in our discussion, we said that we'd like to keep the 

```
  public static readonly LCTRLFLAG = 0x0001;      // Left Control flag
  public static readonly RCTRLFLAG = 0x0002;      // Right Control flag
```

syntax because that's used in Developer. However, `Codes.modifierKeys` which is implemented as 

```
  modifierCodes: {
    "LCTRL":0x0001,           // LCTRLFLAG
    "RCTRL":0x0002,           // RCTRLFLAG
    ...
  } as {[name: string]: number},
```

is also used in Developer (via `MinimalCodesInterface`). The comment in `keyboardHarness.ts` says: 
> Defines any public API points used by debug-compiled keyboards for human-readable
> * nomenclature in rules within a Keyman keyboard's script for keyboard rules.
> *
> * Refer to C:\keymanapp\keyman\developer\src\tike\compile\CompileKeymanWeb.pas,
> * TCompileKeymanWeb.JavaScript_SetupDebug.

Is this still true? Then we will need both?

**Comment by mcdurdin** [comment]
> Is this still true? Then we will need both?

Yes, we'll need both. Keyboards build with Debug configuration have references to `modifierCodes` and `keyCodes`. See how debug-build keyboards use it, for example:

https://gist.github.com/mcdurdin/1aeedf74c2127165bc63f705d41b3292#file-khmer_angkor-js-L9-L10

```js
  var modCodes = keyman.osk.modifierCodes;
  var keyCodes = keyman.osk.keyCodes;
```

**Comment by ermshiperete** [comment]
Superseded by #12072
